File: 19_remove_nth_node_from_end_of_list/removeNthFromEnd.py
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    def getListLength(self, head):
        tmp = head
        cnt = 0
        while tmp:
            cnt += 1
            tmp = tmp.next
        return cnt

    def removeNthFromEnd(self, head, n):
        """
        :type head: ListNode
        :type n: int
        :rtypr: ListNode
        """
        length = self.getListLength(head)
        target = length - n + 1
        if target == 1: return head.next

        #print length, target
        tmp = head
        while target > 2:
            tmp = tmp.next
            target -= 1

        tmp.next = None if not tmp.next.next else tmp.next.next

        return head

if __name__ == '__main__':
    head = ListNode(1)
    #tmp = head
    #tmp.next = ListNode(2)
    #tmp.next.next = ListNode(3)
    n = 1
    
    sol = Solution()
    res = sol.removeNthFromEnd(head, n)
    
    while res:
        print res.val
        res = res.next

File: 19_remove_nth_node_from_end_of_list/removeNthFromEnd.cpp
#include <iostream>

using namespace std;
// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
	public:
    	ListNode* removeNthFromEnd(ListNode* head, int n) {
            int len = getListLength(head);
            int diff = len - n;
            if(!diff) return NULL;
            ListNode *forward = head, *prev = NULL;
            while(diff) {
                prev = forward;
                forward = forward -> next;
                diff--;
            }
            prev -> next = forward -> next;
            delete forward;
            return head;
    	}
    private:
        int getListLength(ListNode* head) {
            ListNode* tmp = head;
            int len = 0;
            while(tmp) {
                len++;
                tmp = tmp -> next;
            }
            return len;
        }
};

int main() {
    ListNode* head = new ListNode(0);
    //head -> next = new ListNode(1);
    //head -> next -> next = new ListNode(2);
    Solution sol;
    ListNode* res = sol.removeNthFromEnd(head, 1);
    while(head) {
        cout << head -> val << " ";
        head = head -> next;
    }
    cout << endl;
}
